<a href="https://colab.research.google.com/github/gcochlar/Bootcamp_DataScience/blob/main/Modulo_04/Bootcamp_Mod4_Aula04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://drive.google.com/uc?id=16O9eMrtSeRDnDpZgmneXj34eIClCntvg'>

# **Tratamento, Análise e** ***Machine Learning*** **Aplicado**

---

Esse notebook servirá para o acompanhamento das aulas do Módulo 04 do ***Bootcamp de Data Science Aplicada*** realizado pela Alura.

Este módulo dará início ao desenvolvimento de um projeto prático que passará por todo o *workflow* em *Data Science*, do entendimento do problema, tratamento e análise dos dados até a proposta de solução utilizando ***Machine Learning***.

Vamos trabalhar com dados da **COVID-19** do **Hospital Sírio Libanês**, focando este módulo no tratamento e análise de dados para entender profundamente o problema que estamos lidando e propor possíveis soluções.

---

## **Aula 01 -** ***Machine Learning*** **e Saúde**
---


In [ ]:
import pandas as pd


In [ ]:
url = 'https://github.com/gcochlar/Bootcamp_DataScience/blob/main/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true'

In [ ]:
dados = pd.read_excel(url)

In [ ]:
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

A primeira aula serviu para explicar o funcionamento do **Kaggle** e nos ajudar a localizar o desafio do **Hospital Sírio Libanês** que está postado lá ([link aqui](https://www.kaggle.com/S%C3%ADrio-Libanes/covid19)).

Também nos mostrou como interpretar inicialmente os dados do desafio e como criar e trabalhar com o notebook diretamente dentro do **Kaggle**.

Além disso, foi explicada a diferença entre ***Algoritmo de Classificação Binária***, que será utilizado nesse desafio e ***Algoritmo de Regressão***.

Na ***Classificação Binária*** buscamos uma resposta *booleana*, o resultado acontece ou não (1 e 0). No presente caso, o paciente vai precisar de internação em UTI ou não.

Na ***Regressão*** buscamos um resultado numa faixa contínua de resultados. Por exemplo, um preço de venda para um apartamento.

Em função da continuidade e para continuar postando os notebooks no **GitHub**, vou continuar usando o **Colab**.

#### **Desafio - Explorar o Kaggle, os datasets, competições e discussões para verificar o quão rico é esta plataforma**
---

Já conhecia o **Kaggle** desde a *Imersão em Data Science* da **Alura**, mas aproveitei a oportunidade para navegar mais um pouco pelo site.

Achei interessante a disponibilidade de bases de dados de assuntos diversos como cinema, quadrinhos e música, além de assuntos mais sérios como já era esperado.

Não tinha notado também que o **Kaggle** oferece alguns cursos, pretendo retornar posteriormente para ver o que eles disponibilizam em termos de ***Machine Learning***.

#### **Desafio - Leia a descrição do problema proposta pelo Sírio Libanês, tentando entender qual é o real problema. Antes de continuar para as próximas aulas, debata no Discord suas conclusões sobre o projeto**
---

O problema proposto no desafio estabelece duas tarefas diferentes. Vi que no **Discord** houve uma certa discussão a respeito de serem a mesma coisa ou não, mas o meu entendimento é de que são tarefas diferentes mesmo.

>**Task 01**
>
>Predict admission to the ICU of confirmed COVID-19 cases.
>
>Based on the data available, is it feasible to predict which patients will need intensive care unit support?
>
>The aim is to provide tertiary and quarternary hospitals with the most accurate answer, so ICU resources can be arranged or patient transfer can be scheduled.

Essa tarefa busca maximizar o acerto de casos positivos minimizando os falsos negativos, ou seja, acertar o máximo de casos de pacientes que irão precisar de UTI, para evitar o sub- ou superdimensionamento da estrutura.

>**Task 02**
>
>Predict NOT admission to the ICU of confirmed COVID-19 cases.
>
>Based on the subsample of widely available data, is it feasible to predict which patients will need intensive care unit support?
>
>The aim is to provide local and temporary hospitals a good enough answer, so frontline physicians can safely discharge and remotely follow up with these patients.

Essa tarefa busca maximizar o acerto de casos negativos minimizando os falsos positivos, ou seja, acertar o máximo de casos que não irão precisar de UTI e que poderiam receber alta e ser monitorados à distância, liberando a estrutura hospitalar para receber mais pacientes.

### **O que aprendemos nessa aula?**
---

>* O que é o **Kaggle** e como utilizar;
* Como interpretar o problema proposto pelo **Sírio Libanês**;
* O que é uma Classificação Binária.

---


## **Aula 02 - Desenvolvimento no** ***Kaggle***
---


Apesar da aula continuar usando o **Kaggle**, vou permanecer no **Colab**.

Vamos começar analisando o que temos na base de dados importada.

In [ ]:
dados.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

In [ ]:
dados['ICU'].value_counts()

0    1410
1     515
Name: ICU, dtype: int64

#### **Desafio - Pesquisar Aprendizado Supervisionado, Não Supervisionado, Por Reforço e Mais**
---

**Aprendizado Supervisionado**

É o modelo mais comum. Treinamos o algoritmo com um volume inicial de dados, mostrando características e o resultado obtido. O processo de aprendizado leva a prever a saída a partir de novos dados de entrada.

**Aprendizado Não Supervisionada**

É o modelo utilizado quando não temos os dados de resultado obtido na base de dados inicial. O algoritmo tenta entender as características e gerar agrupamentos (*clusters*), muitas vezes para depois passar cada *cluster* por um processo de aprendizado supervisionado.

**Aprendizado Por Reforço**

É o modelo mais indicado para tomada de decisões. Funciona por tentativa e erro, onde a máquina é exposta a vários processos decisórios e vai aprendendo conforme as decisões tomadas forem acertadas ou não.

**Redes Neurais e** ***Deep Learning***

Trabalha com sucessivas camadas de aprendizado para melhorar o processo de tomada de decisão. É utilizado para descobrir padrões em dados desestruturados, ideal para reconhecimento de imagens e de fala, por exemplo.

Normalmente as camadas são formadas por combinações de algoritmos de aprendizado supervisionado e não supervisionado.

Fontes:

[Data Hackers](https://medium.com/data-hackers/simplificando-as-categorias-de-algoritmos-de-machine-learning-191ac9dc8ddf)

[IBM - Machine Learning for Dummies](https://www.ibm.com/downloads/cas/GB8ZMQZ3)

[LAMFO - UNB](https://lamfo-unb.github.io/2017/07/27/tres-tipos-am/)


#### **Desafio - Pesquisar Sobre Regressão, Classificação, etc.**
---

A diferença entre a **Regressão** e a **Classificação** se dá normalmente pelo tipo de resultado que teremos.

Para modelos de **Classificação** teremos um número limitado e discreto de resultados.

Já para modelos de **Regressão** o resultado esperado está numa faixa contínua de dados. É mais utilizado para encontrar a correlação entre variáveis.

### **O que aprendemos nessa aula?**
---

>* Como usar os notebooks direto do **Kaggle**;
>
>* O que são algoritmos supervisionados;
>
>* O que são dados anonimizados e sua relevância da perspectiva ética.

---


## **Aula 03 - Modelos de** ***Machine Learning***
---


Para iniciar o nosso estudo, vamos usar a biblioteca **```SciKitLearn```** e o seu modelo de Regressão Logística.

In [ ]:
from sklearn.linear_model import LogisticRegression

modelo = LogisticRegression()

modelo

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Vamos preparar a nossa base de dados para manter só as colunas com números, deixar de fora as linhas com dados nulos e NaN e também a coluna 'ICU', que é a coluna que contém o resultado.

Essa coluna 'ICU' vai para uma tabela em separado, para avaliarmos o resultado depois.

In [ ]:
# Pegar só as colunas que contém números
x_columns = dados.describe().columns

# Excluir linhas que contém NaN
dados_limpos = dados.dropna()

# Separar os dados de resultado (y) e de entrada (x)
y = dados_limpos.ICU
x = dados_limpos[x_columns].drop(['ICU'], axis = 1)

In [ ]:
modelo.fit(x,y)
prev = modelo.predict(x)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [ ]:
print(sum(prev == y),'acertos')
print(sum(prev != y),'erros')
print(sum(prev == y) / len(x) * 100,'% de acertos')
print(sum(prev != y) / len(x) * 100,'% de erros')

496 acertos
49 erros
91.00917431192661 % de acertos
8.990825688073395 % de erros


#### **Desafio - Pesquisar Sobre Ética em** ***Data Science*** **e Inteligência Artificial**
---



https://insider.dn.pt/em-rede/privacidade-no-data-science-a-etica-deve-ir-muito-alem-da-lei/25736/

https://www.serpro.gov.br/lgpd/noticias/2019/tratamento-etico-dados-pessoais-ciencia-data-science

http://www.sciencefriday.com/articles/10-questions-for-the-nations-first-chief-data-scientist/

https://medium.com/somos-tera/o-que-e-etica-de-dados-3ce4702cdfc1

https://royalsocietypublishing.org/doi/full/10.1098/rsta.2016.0360#sec-1

https://www.itinsight.pt/news/data-and-analytics/insights-2018---etica-estrategia-e-futuro-da-data-science-em-debate

https://neofeed.com.br/blog/home/qual-e-a-etica-da-inteligencia-artificial/

https://pt.wikipedia.org/wiki/%C3%89tica_na_intelig%C3%AAncia_artificial

https://www.umov.me/inteligencia-artificial-etica-e-trabalho/



#### **Desafio - Qual o motivo de jogar fora o ICU dos dados X? O que acontece se estiver nos dados X? Isso seria bom ou ruim?**
---



Como podemos ver no resultado do código abaixo, se rodarmos o modelo com os dados da coluna 'ICU' inclusas no parâmetro X, o modelo será treinado para responder com 100% de acerto para esses dados de treino, pois ele acaba aprendendo que sempre que receber no *input* que 'ICU' = 1, ele vai responder que o *output*, que também é 'ICU' será 1, o que é óbvio.

Mas esse modelo não vai ter nenhuma utilidade, pois vai depender de termos a resposta da coluna 'ICU' para nos dar a previsão de resposta (que já teremos). Passa a ser um modelo de previsão do passado.

In [ ]:
modelo2 = LogisticRegression();
x2 = dados_limpos[x_columns]
modelo2.fit(x2,y);
print(sum(modelo2.predict(x2) == y),'acertos');
print(sum(modelo2.predict(x2) != y),'erros');
print(sum(modelo2.predict(x2) == y) / len(x2) * 100,'% de acertos');
print(sum(modelo2.predict(x2) != y) / len(x2) * 100,'% de erros');

545 acertos
0 erros
100.0 % de acertos
0.0 % de erros


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### **O que aprendemos nessa aula?**
---

>* O que é a uma regressão logística;
>
>* Como treinar o seu primeiro modelo de ***Machine Learning***.

---


## **Aula 04 - Métricas e Avaliações**
---


Vamos melhorar o nosso modelo, mas antes vamos ver qual seria a classificação de um modelo **```DummyClassifier```**, que simplesmente classifica tudo com base na proporção que temos na base de dados.

Vamos verificar qual a proporção da base de dados.

In [ ]:
y.value_counts()

0    281
1    264
Name: ICU, dtype: int64

Vemos que o resultado 0 está em 281 linhas, ou seja, 51,56% da base.

In [ ]:
from sklearn.dummy import DummyClassifier

modelo_dummy = DummyClassifier()
modelo_dummy.fit(x,y)
prev_dummy = modelo_dummy.predict(x)
print(sum(prev_dummy == y) / len(x) * 100,'% de acertos');

51.559633027522935 % de acertos


/usr/local/lib/python3.7/dist-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


In [ ]:
import numpy as np

teste = np.bincount(prev_dummy) ## Equivalente ao value_counts para arrays
ii = np.nonzero(teste)[0]

np.vstack((ii,teste[ii])).T

array([[  0, 289],
       [  1, 256]])

A biblioteca **```sklearn```** também tem uma ferramenta de avaliação de acurácia do modelo, para nos permitir comparações.

No presente caso, a resposta vai ser a mesma que já calculamos acima.

In [ ]:
from sklearn.metrics import accuracy_score

accuracy_score(y, prev_dummy)

0.5155963302752293

In [ ]:
accuracy_score(y, modelo2.predict(x2))

1.0

In [ ]:
accuracy_score(y, prev)

0.9100917431192661

O problema de fazermos o treino com toda a base é que isso aumenta artificialmente a acurácia do modelo, pois ele vai ser testado contra dados que também foram usados no treino.

Para que o modelo esteja preparado para trabalhar com dados novos, precisamos testar a acurácia contra dados que não foram usados no treino.

Para isso, vamos agora fazer a separação dos nossos dados em treino e teste, para poder criar um modelo e depois verificar a acurácia contra dados que ele nunca viu antes.

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y)

In [ ]:
print(len(x_train), len(y_train))
print(len(x_test), len(y_test))

408 408
137 137


In [ ]:
modelo_2 = LogisticRegression()
modelo_2.fit(x_train, y_train)

y_prev = modelo_2.predict(x_test)
accuracy_score(y_test, y_prev)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.9051094890510949

In [ ]:
np.random.seed(73246)

x3_train, x3_test, y3_train, y3_test = train_test_split(x, y, stratify = y)
print(len(x3_train), len(y3_train))
print(len(x3_test), len(y3_test))

408 408
137 137


In [ ]:
modelo_3 = LogisticRegression()
modelo_3.fit(x3_train, y3_train)

y3_prev = modelo_3.predict(x3_test)
accuracy_score(y3_test, y3_prev)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8321167883211679

#### **Desafio - Decida quanto você vai escolher para treino e quanto para teste? Quanto as pessoas utilizam e qual o motivo?**
---

A divisão mais comum é na proporção 70%\~30% entre dados de treino e de teste. Gosto de usar essa proporção sempre que possível. No caso específico do COVID-19 tenho encontrado dificuldade para utilizar essa proporção em razão da pequena quantidade de dados disponíveis.

No módulo anterior foi utilizado quase 84% (400 das 478 linhas) dos dados para treino, justamente por causa disso.

A função **```train_test_split```** faz a divisão tendo como parâmetro *default* a proporção de 75%\~25%, mas isso pode ser alterado mexendo no parâmetro **```test_size```**, que recebe um número entre 0 e 1 que vai indicar a proporção do teste (o resto é treino).

Para fazer uma divisão entre dados de treino, de teste e de validação, encontrei a função abaixo.


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

def split_stratified_into_train_val_test(df_input, stratify_colname='y',
                                         frac_train=0.6, frac_val=0.15, frac_test=0.25,
                                         random_state=None):
    '''
    Splits a Pandas dataframe into three subsets (train, val, and test)
    following fractional ratios provided by the user, where each subset is
    stratified by the values in a specific column (that is, each subset has
    the same relative frequency of the values in the column). It performs this
    splitting by running train_test_split() twice.

    Parameters
    ----------
    df_input : Pandas dataframe
        Input dataframe to be split.
    stratify_colname : str
        The name of the column that will be used for stratification. Usually
        this column would be for the label.
    frac_train : float
    frac_val   : float
    frac_test  : float
        The ratios with which the dataframe will be split into train, val, and
        test data. The values should be expressed as float fractions and should
        sum to 1.0.
    random_state : int, None, or RandomStateInstance
        Value to be passed to train_test_split().

    Returns
    -------
    df_train, df_val, df_test :
        Dataframes containing the three splits.
    '''

    if frac_train + frac_val + frac_test != 1.0:
        raise ValueError('fractions %f, %f, %f do not add up to 1.0' % \
                         (frac_train, frac_val, frac_test))

    if stratify_colname not in df_input.columns:
        raise ValueError('%s is not a column in the dataframe' % (stratify_colname))

    X = df_input # Contains all columns.
    y = df_input[[stratify_colname]] # Dataframe of just the column on which to stratify.

    # Split original dataframe into train and temp dataframes.
    df_train, df_temp, y_train, y_temp = train_test_split(X,
                                                          y,
                                                          stratify=y,
                                                          test_size=(1.0 - frac_train),
                                                          random_state=random_state)

    # Split the temp dataframe into val and test dataframes.
    relative_frac_test = frac_test / (frac_val + frac_test)
    df_val, df_test, y_val, y_test = train_test_split(df_temp,
                                                      y_temp,
                                                      stratify=y_temp,
                                                      test_size=relative_frac_test,
                                                      random_state=random_state)

    assert len(df_input) == len(df_train) + len(df_val) + len(df_test)

    return df_train, df_val, df_test

#### **Desafio - Olhe a documentação da ```LogisticRegression```... o que tem de parâmetro em comum com o ```train_test_split``` e o que isso significa para nós?**
---

O parâmetro que ambos têm em comum é o **```random_state```**.

Acredito que sirva para garantir a reprodutibilidade do estudo, mas no caso da Regressão Logística, esse parâmetro só tem efeito para alguns algoritmos (e o *default* não está entre eles).

#### **OBS.:** Não consegui atingir o mesmo resultado que o vídeo da aula.

Copiei o código da aula disponibilizado abaixo, idêntico, em uma célula única, abaixo. Mesmo assim, não fecha o resultado, não entendi...

In [ ]:
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

import pandas as pd

dados = pd.read_excel("https://github.com/gcochlar/Bootcamp_DataScience/blob/main/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true")
dados.head()

dados.describe()

dados["ICU"]

dados["ICU"].value_counts()

x_columns = dados.describe().columns

dados_limpos = dados.dropna()

y = dados_limpos["ICU"]
x = dados_limpos[x_columns].drop(["ICU"], axis=1)

from sklearn.linear_model import LogisticRegression

modelo = LogisticRegression()
modelo.fit(x, y)

modelo.predict([x.iloc[4]])

x.iloc[4]

y.iloc[4]

sum(modelo.predict(x) == y)

sum(modelo.predict(x) != y)

accuracy = sum(modelo.predict(x) == y) / len(y) * 100
accuracy

y.value_counts()

from sklearn.dummy import DummyClassifier

modelo = DummyClassifier("most_frequent")
modelo.fit(x, y)
sum(modelo.predict(x) == y) / len(y) * 100

from sklearn.metrics import accuracy_score

accuracy_score(y, modelo.predict(x))

import numpy as np

np.random.seed(73246)

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

print(len(x_train), len(y_train))
print(len(x_test), len(y_test))

modelo = LogisticRegression()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)


408 408
137 137


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


0.8321167883211679

No *notebook* que tem só o conteúdo da aula, o resultado é 0.8248175182481752.

A mesma célula, copiada para cá, dá esse resultado acima (0.8321167883211679). Isso não tem explicação...

### **O que aprendemos nessa aula?**
---

>* Como metrificar um modelos de classificação, usando o **```Sci-kit Learn```**;
>
>* Como realizar o treinamento dividindo em dados de treino e teste.

---
